doc/yarp_buffering.dox

```diff
@@ -0,0 +1,83 @@
+
+Methods that can be useful to monitor the status of read and write operations are: yarp::os::BufferedPort::getPendingReads() and yarp::os::BufferedPort::isWriting().
+
+It is also important to understand that a BufferedPort is managing the life-cycle of the messages transmitted on the network. This means that the BufferedPort will allocate and re-cycle objects in its internal buffers.
+
+\section s_buffering_writing Life-cycle: writing to a port
+
+When you call yarp::os::BufferedPort::prepare() you get a slot from the internal buffer. This slot is reserved for you, until the next call to BufferedPort::write(). This means that you are free to write inside the object returned by BufferedPort::prepare(), until the next call to BufferedPort::write(). At this point the YARP infrastructure will begin transmitting the message and you are no longer allowed to modify it (well, actually nobody forbids doing so, but in this case you may have unexpected effects, i.e. probably the message will be corrupted).
+
+Objects are recycled within the BufferedPort to save CPU cycles, so the next time you call yarp::os::BufferedPort::prepare() you will not get an empty object. Make sure you clean the object before you write to it (for example do not append data to a Bottle), otherwise you will get unexpected behavior (likely the object will contain past data). This code example illustrates this concept:
+
+\code 
+ BufferedPort<Bottle> port;
+ ...
+ 
+ Bottle &b=port.prepare();
+ //write to b
+ port.write();
+ // don't touch b!
+\endcode
+
+\section  s_buffering_reading Life-cycle: reading from a port
+To read from a BufferedPort you call yarp::os::BufferedPort::read(), this returns a pointer to the message which arrived most recently (or the oldest message, if you changed the buffering policy to FIFO using yarp::os::BufferedPort::setStrict()). This object is reserved and not touched by YARP, until the next call to yarp::os::BufferedPort::read(). This code example clarify this issue:
+
+\code 
+ BufferedPort<Bottle> port;
+ ...
+ 
+ Bottle *b1=port.read();
+ // read b1, YARP will not touch it until the next call to read().
+ 
+ Bottle *b2=port.read();
+ // now read b2, YARP will not touch it until the next call to read().
+ // The object pointed to by b1 can be re-cycled by YARP to store new data. 
+ // If you access b1 you may get garbage. 
+\endcode
+
+In practice this is a mistake people rarely do, because it is much better to read messages from a Port in a loop, e.g.:
+
+\code 
+ BufferedPort<Bottle> port;
+ ...
+ while (true) {
+    Bottle *b1=port.read();
+    // read b1, YARP will not touch it until the next call to read().
+ 
+ }
+  
+\endcode
+
+however, it is important to keep in mind.
```

Comment:
@lornat75, something like this?

```
During common practice, BufferPorts are used within a loop to read data, like e.g.
\code 
BufferedPort<Bottle> port;
...
while (true) {
Bottle *b1=port.read();
// read b1, YARP will not touch it until the next call to read().
}
\endcode
which inherently avoids the previously described problem of accessing recycled data.
However, it is important to keep in mind the data life-cycle inside a BufferPort to avoid undefined behaviors.
```